### ISSUE TITLE
[as3/starling] Unexpected number of draw calls

### ISSUE THREAD

**Issue by badlogic** [EsotericSoftware/spine-runtimes#1077]
See http://esotericsoftware.com/forum/AS3-Starling-Performance-Optimization-9914

**Comment by badlogic** [comment]
Can not reproduce. Using test skeleton by user and the following code:

```
/******************************************************************************
 * Spine Runtimes Software License v2.5
 *
 * Copyright (c) 2013-2016, Esoteric Software
 * All rights reserved.
 *
 * You are granted a perpetual, non-exclusive, non-sublicensable, and
 * non-transferable license to use, install, execute, and perform the Spine
 * Runtimes software and derivative works solely for personal or internal
 * use. Without the written permission of Esoteric Software (see Section 2 of
 * the Spine Software License Agreement), you may not (a) modify, translate,
 * adapt, or develop new applications using the Spine Runtimes or otherwise
 * create derivative works or improvements of the Spine Runtimes or (b) remove,
 * delete, alter, or obscure any trademarks or any copyright, trademark, patent,
 * or other intellectual property or proprietary rights notices on or in the
 * Software, including any copy thereof. Redistributions in binary or source
 * form must include this license and terms.
 *
 * THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL ESOTERIC SOFTWARE BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES, BUSINESS INTERRUPTION, OR LOSS OF
 * USE, DATA, OR PROFITS) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

package spine.examples {
	import spine.*;
	import spine.atlas.Atlas;
	import spine.attachments.AtlasAttachmentLoader;
	import spine.attachments.AttachmentLoader;
	import spine.starling.SkeletonAnimation;
	import spine.starling.StarlingTextureLoader;

	import starling.core.Starling;
	import starling.display.Sprite;

	public class TestScene extends Sprite {
		[Embed(source = "/skeleton.json", mimeType = "application/octet-stream")]
		static public const TestJson : Class;

		[Embed(source = "/skeleton.atlas", mimeType = "application/octet-stream")]
		static public const TestAtlas : Class;

		[Embed(source = "/skeleton.png")]
		static public const TestTexture : Class;		

		public function TestScene() {
			var attachmentLoader : AttachmentLoader;
			var spineAtlas : Atlas = new Atlas(new TestAtlas(), new StarlingTextureLoader(new TestTexture()));
			attachmentLoader = new AtlasAttachmentLoader(spineAtlas);

			var json : SkeletonJson = new SkeletonJson(attachmentLoader);
			json.scale = 0.5;
			var skeletonData : SkeletonData = json.readSkeletonData(new TestJson());

			for (var i: int = 0; i < 10; i++) {
				var skeleton : SkeletonAnimation = new SkeletonAnimation(skeletonData);
				skeleton.x = Math.random() * 400 + 200;
				skeleton.y = Math.random() * 400 + 200;
				skeleton.state.setAnimationByName(0, "animation", true);
				addChild(skeleton);
			}
			
			Starling.juggler.add(skeleton);			
		}
	}
}
```

I get the expected 1 draw call:

![https://sht.tl/CIbLtZ](https://sht.tl/CIbLtZ)

@neuronix I'll need a better test sample to reproduce your error. However, I've checked the code for any obvious batch breaking issues. Everything works as intended. One thing that can break your batching is our use of a custom mesh style and effect for 2 color tinting.

**Comment by neuronix** [comment]
Hi @badlogic, I'm really sorry for the false alarm.. my custom viewer was creating extra draw calls when listing the animations.. got a bit of tunnel vision there. Thanks for checking it out. 

Bad news for us is that we'll have to look for other optimizations to improve performance on Android.

**Comment by badlogic** [comment]
@neuronix no, there might actually not be bad news. I've changed the default behaviour of `SkeletonSprite`. Previously, by default, `SkeletonSprite` uses its own vertex layout, mesh style and effect. If you intersperse rendering of `SkeletonSprite` with say Starling's `Image`, then that would break batching and result in more draw calls.

With the latest change (available in the 3.6 branch), by default, `SkeletonSprite` will use the default vertex layout, mesh style and effect. When mixing rendering of `SkeletonSprite` with other Starling objects, there shouldn't be any batch breaking (as long as they use the same texture and blending), and thus reduce draw calls.

Update to the latest code and give it a try.

**Comment by neuronix** [comment]
Thank you @badlogic, I will test this :)
